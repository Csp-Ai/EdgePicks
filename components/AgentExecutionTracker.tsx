import React, { useEffect, useMemo, useRef, useState } from 'react';
import { CheckCircle2, AlertTriangle, Circle } from 'lucide-react';

export type AgentStatus = 'pending' | 'running' | 'completed' | 'error';

export interface AgentMeta {
  name: string;
  label?: string;
  icon?: React.ReactNode;
}

export interface LifecycleEvent {
  name: string;
  status: AgentStatus;
  timestamp?: number;
}

interface Props {
  agents: AgentMeta[];
  events?: LifecycleEvent[];
  mode?: 'live' | 'demo';
}

const statusRank: Record<AgentStatus, number> = {
  pending: 0,
  running: 1,
  completed: 2,
  error: 2,
};

const mergeStatus = (prev: AgentStatus, next: AgentStatus): AgentStatus => {
  return statusRank[next] >= statusRank[prev] ? next : prev;
};

const AgentExecutionTracker: React.FC<Props> = ({
  agents,
  events = [],
  mode = 'live',
}) => {
  const [runId, setRunId] = useState(0);
  const base = useMemo(
    () => Object.fromEntries(agents.map((a) => [a.name, 'pending' as AgentStatus])),
    [agents, runId],
  );

  const [statuses, setStatuses] = useState<Record<string, AgentStatus>>(base);
  const timeoutsRef = useRef<Array<ReturnType<typeof setTimeout>>>([]);
  const subscriptionsRef = useRef<Array<() => void>>([]);

  // cleanup on unmount
  useEffect(() => {
    return () => {
      timeoutsRef.current.forEach(clearTimeout);
      timeoutsRef.current = [];
      subscriptionsRef.current.forEach((unsub) => unsub());
      subscriptionsRef.current = [];
    };
  }, []);

  // apply events
  useEffect(() => {
    setStatuses((prev) => {
      const next = { ...base };
      for (const e of events) {
        if (!(e.name in next)) continue;
        next[e.name] = mergeStatus(next[e.name], e.status);
      }
      return next;
    });
  }, [events, base]);

  // demo mode simulation
  useEffect(() => {
    if (mode !== 'demo') return;
    agents.forEach((agent, i) => {
      const start = setTimeout(() => {
        setStatuses((s) => ({ ...s, [agent.name]: 'running' }));
      }, i * 1000);
      const end = setTimeout(() => {
        setStatuses((s) => ({ ...s, [agent.name]: 'completed' }));
      }, i * 1000 + 800);
      timeoutsRef.current.push(start, end);
    });
    return () => {
      timeoutsRef.current.forEach(clearTimeout);
      timeoutsRef.current = [];
    };
  }, [mode, agents, runId]);

  const allDone = useMemo(
    () =>
      agents.every(
        (a) => statuses[a.name] === 'completed' || statuses[a.name] === 'error',
      ),
    [agents, statuses],
  );

  const replay = () => {
    setRunId((r) => r + 1);
    setStatuses(base);
  };

  return (
    <div className="w-full text-white" data-testid="agent-tracker">
      <div className="text-center mb-4">
        <h3 className="text-lg font-semibold">Live Agent Execution</h3>
        <p className="text-xs text-slate-400">Powered by Modular AI Agents</p>
      </div>
      <div className="flex items-center justify-center gap-4 flex-wrap">
        {agents.map((agent, idx) => {
          const status = statuses[agent.name];
          const Icon =
            status === 'completed'
              ? CheckCircle2
              : status === 'error'
              ? AlertTriangle
              : Circle;
          return (
            <div key={agent.name} className="flex flex-col items-center">
              <div
                data-testid={`node-${agent.name}`}
                data-status={status}
                className={`relative w-12 h-12 flex items-center justify-center rounded-full border-2 transition-colors ${
                  status === 'running'
                    ? 'border-blue-500 animate-pulse shadow-[0_0_8px_rgba(59,130,246,0.7)]'
                    : status === 'completed'
                    ? 'border-green-500'
                    : status === 'error'
                    ? 'border-red-500'
                    : 'border-slate-600'
                }`}
              >
                {agent.icon || <Icon className="w-6 h-6" />}
              </div>
              <span className="mt-2 text-xs text-center w-16 truncate">
                {agent.label || agent.name}
              </span>
              {idx < agents.length - 1 && (
                <span className="hidden sm:block absolute top-6 left-full w-8 h-px bg-slate-600" />
              )}
            </div>
          );
        })}
      </div>
      {allDone && (
        <div className="text-center mt-4" data-testid="flow-complete">
          <div className="inline-block px-3 py-1 bg-green-600/20 text-green-400 rounded">
            Flow Complete
          </div>
          {mode === 'demo' && (
            <div>
              <button
                className="mt-2 text-sm underline text-blue-400"
                onClick={replay}
              >
                Replay
              </button>
            </div>
          )}
        </div>
      )}
    </div>
  );
};

export default AgentExecutionTracker;

