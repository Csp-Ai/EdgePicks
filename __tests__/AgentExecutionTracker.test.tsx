import React from 'react';
import { render, screen, act } from '@testing-library/react';
import AgentExecutionTracker, {
  LifecycleEvent,
  AgentMeta,
} from '../components/AgentExecutionTracker';

const agents: AgentMeta[] = [
  { name: 'injuryScout', label: 'Injury' },
  { name: 'lineWatcher', label: 'Lines' },
];

describe('AgentExecutionTracker', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.runOnlyPendingTimers();
    jest.clearAllTimers();
    jest.useRealTimers();
  });

  it('updates status based on events and handles out-of-order and missing agents', () => {
    const { rerender } = render(
      <AgentExecutionTracker agents={agents} events={[]} mode="live" />,
    );
    expect(screen.getByTestId('node-injuryScout')).toHaveAttribute(
      'data-status',
      'pending',
    );

    const events1: LifecycleEvent[] = [
      { name: 'injuryScout', status: 'completed' },
      { name: 'lineWatcher', status: 'running' },
      { name: 'unknown', status: 'running' },
    ];
    rerender(
      <AgentExecutionTracker agents={agents} events={events1} mode="live" />,
    );
    expect(screen.getByTestId('node-injuryScout')).toHaveAttribute(
      'data-status',
      'completed',
    );
    expect(screen.getByTestId('node-lineWatcher')).toHaveAttribute(
      'data-status',
      'running',
    );

    const events2: LifecycleEvent[] = [
      ...events1,
      { name: 'injuryScout', status: 'running' }, // out of order
      { name: 'lineWatcher', status: 'error' },
    ];
    rerender(
      <AgentExecutionTracker agents={agents} events={events2} mode="live" />,
    );
    expect(screen.getByTestId('node-injuryScout')).toHaveAttribute(
      'data-status',
      'completed',
    );
    expect(screen.getByTestId('node-lineWatcher')).toHaveAttribute(
      'data-status',
      'error',
    );
  });

  it('completes flow in demo mode', async () => {
    render(<AgentExecutionTracker agents={agents} events={[]} mode="demo" />);
    await act(async () => {
      jest.advanceTimersByTime(3000);
    });
    await screen.findByTestId('flow-complete');
  });

  it('matches initial snapshot', () => {
    const { container } = render(
      <AgentExecutionTracker agents={agents} events={[]} mode="live" />,
    );
    expect(container).toMatchSnapshot();
  });

  it('matches final snapshot', () => {
    const { container } = render(
      <AgentExecutionTracker
        agents={agents}
        events={[
          { name: 'injuryScout', status: 'completed' },
          { name: 'lineWatcher', status: 'completed' },
        ]}
        mode="live"
      />,
    );
    expect(container).toMatchSnapshot();
  });
});

