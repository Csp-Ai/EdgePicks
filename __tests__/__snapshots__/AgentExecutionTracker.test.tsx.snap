// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AgentExecutionTracker matches final snapshot 1`] = `
<div>
  <div
    class="w-full text-white"
    data-testid="agent-tracker"
  >
    <div
      class="text-center mb-4"
    >
      <h3
        class="text-lg font-semibold"
      >
        Live Agent Execution
      </h3>
      <p
        class="text-xs text-slate-400"
      >
        Powered by Modular AI Agents
      </p>
    </div>
    <div
      class="flex items-center justify-center gap-4 flex-wrap"
    >
      <div
        class="flex flex-col items-center"
      >
        <div
          class="relative w-12 h-12 flex items-center justify-center rounded-full border-2 transition-colors border-green-500"
          data-status="completed"
          data-testid="node-injuryScout"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-circle-check w-6 h-6"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="12"
              cy="12"
              r="10"
            />
            <path
              d="m9 12 2 2 4-4"
            />
          </svg>
        </div>
        <span
          class="mt-2 text-xs text-center w-16 truncate"
        >
          Injury
        </span>
        <span
          class="hidden sm:block absolute top-6 left-full w-8 h-px bg-slate-600"
        />
      </div>
      <div
        class="flex flex-col items-center"
      >
        <div
          class="relative w-12 h-12 flex items-center justify-center rounded-full border-2 transition-colors border-green-500"
          data-status="completed"
          data-testid="node-lineWatcher"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-circle-check w-6 h-6"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="12"
              cy="12"
              r="10"
            />
            <path
              d="m9 12 2 2 4-4"
            />
          </svg>
        </div>
        <span
          class="mt-2 text-xs text-center w-16 truncate"
        >
          Lines
        </span>
      </div>
    </div>
    <div
      class="text-center mt-4"
      data-testid="flow-complete"
    >
      <div
        class="inline-block px-3 py-1 bg-green-600/20 text-green-400 rounded"
      >
        Flow Complete
      </div>
    </div>
  </div>
</div>
`;

exports[`AgentExecutionTracker matches initial snapshot 1`] = `
<div>
  <div
    class="w-full text-white"
    data-testid="agent-tracker"
  >
    <div
      class="text-center mb-4"
    >
      <h3
        class="text-lg font-semibold"
      >
        Live Agent Execution
      </h3>
      <p
        class="text-xs text-slate-400"
      >
        Powered by Modular AI Agents
      </p>
    </div>
    <div
      class="flex items-center justify-center gap-4 flex-wrap"
    >
      <div
        class="flex flex-col items-center"
      >
        <div
          class="relative w-12 h-12 flex items-center justify-center rounded-full border-2 transition-colors border-slate-600"
          data-status="pending"
          data-testid="node-injuryScout"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-circle w-6 h-6"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="12"
              cy="12"
              r="10"
            />
          </svg>
        </div>
        <span
          class="mt-2 text-xs text-center w-16 truncate"
        >
          Injury
        </span>
        <span
          class="hidden sm:block absolute top-6 left-full w-8 h-px bg-slate-600"
        />
      </div>
      <div
        class="flex flex-col items-center"
      >
        <div
          class="relative w-12 h-12 flex items-center justify-center rounded-full border-2 transition-colors border-slate-600"
          data-status="pending"
          data-testid="node-lineWatcher"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-circle w-6 h-6"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="12"
              cy="12"
              r="10"
            />
          </svg>
        </div>
        <span
          class="mt-2 text-xs text-center w-16 truncate"
        >
          Lines
        </span>
      </div>
    </div>
  </div>
</div>
`;
